### ISSUE TITLE
[Bug]:The parameter gpu_memory_utilization does not take effect

### ISSUE THREAD

**Issue by liutao053877** [vllm-project/vllm#10637]
### Your current environment

<details>
<summary>The output of `python collect_env.py`</summary>

```text
Your output of `python collect_env.py` here
```

</details>

### Model Input Dumps

Google Colab with T4 GPU

### 🐛 Describe the bug

I set gpu_memory_utilization=0.9, but I find the memory used rate is 0.99 sometimes . It is higher than the value I set before and it sounds like a bug .

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by DarkLight1337** [comment]
Can you show more details, such as which model you are using?

**Comment by DarkLight1337** [comment]
It might be that 90% GPU isn't enough for your model by default. Can you try reducing the memory usage, such as by reducing `max_model_len` and/or `max_num_seqs`?

**Comment by nFunctor** [comment]
Can this be somehow related to Marlin kernels ? I have still the issue #8853 open. I keep using Llama-3.1-70B-Instruct-AWQ-INT4 on L40S and periodic outofmem still occur even after the server launched completely and treated a selection of requests. It feels like a memory leak.

I have experimented with preemption mode and block size, setting the former to swap and the latter to 32. It seemed to have worked for more stability, but since I haven't found a reliable way to reproduce the crash, I cannot confirm for 100%.

**Comment by DarkLight1337** [comment]
cc @joerunde @tlrmchlsmth

**Comment by joerunde** [comment]
Can y'all try again with v0.6.5? The change from #10511 fixed more issues with the memory profiling. It's all still an estimation though, it's expected that the memory used will go over the limit, which is why we don't default it to `.99` or `1`

**Comment by StrikerRUS** [comment]
@joerunde I'm sorry, but I personally won't be able to check new release till mid-January. Maybe anyone else could do it and report right now.

However, your statement "it's expected that the memory used will go over the limit" sounds like a clear bug in memory profiling or in some other vLLM process,  but definitely not like something that is "expected" 😐

**Comment by nFunctor** [comment]
@joerunde the AWQ setup from the issue that I referenced seems to be a bit more stable. I've still had crashes with it and other setups but they happened upfront on the arrival of the first call, which then passed once I reduced gpu mem usage. Since I'm using a lot of prefix caching, I wonder if the evictor patch also helped.

Overall still not completely clear but feels much better, thanks!

**Comment by joerunde** [comment]
> However, your statement "it's expected that the memory used will go over the limit" sounds like a clear bug in memory profiling or in some other vLLM process, but definitely not like something that is "expected" 😐

@StrikerRUS Heh, knowing in advance how much memory a program will take without running it first is hard! The memory profiling runs before the KV cache is allocated or the cuda graphs are captured, and we try to send in a bunch of dummy data where we take the model's maximum context length and divide that up over a batch of size `--max-num-seqs` and run a couple forward passes to see how much memory is used to run the model. This gets us pretty close, but:
- While `torch` provides an api to track peak memory usage, we don't have a way to track peak vram allocations outside of torch
- We don't know how much vram the cuda graphs will take up ahead of time
- There are inputs that users can give like requesting prompt logprobs on long prompts that may take more vram than we see during profiling
- We're not trying to actually manage memory allocation on the devices (we leave that up to torch, etc) so we don't account for things like the extra space taken up by fragmented memory blocks

The team has been putting a lot of work into it and it's getting better but there's still more to implement and we welcome any feedback and help implementing better solutions. For your specific situation, those extra 600MiB or so that are showing as allocated could be due to:
- cuda graphs
- fragmentation
- memory that hasn't been GCed from running requests
- memory that was used on the device outside of vLLM


> I'm running into this using llama-3.2-vision-instruct. I can perfectly fine load the model on the same machine using transformers (not using quantization). I'm doing offline inference.

> When using vllm I'm used to if I set a small value gpu_memory_utilization=0.3 to also see vllm pre-allocate that memory before loading the model. This is somehow also not happening so I assume my parameter is not respected when loading a model as

@chris-aeviator That's fully expected behavior, the profiling code accounts for the vram needed to run requests and will pre-allocate a smaller kv cache so that the memory remains available for the model to use within that limit of `gpu_memory_utilization=0.3`. For example on an 80GB A100:

```
>>> model = LLM("meta-llama/Llama-3.2-11B-Vision-Instruct", gpu_memory_utilization=0.6, max_num_seqs=32, max_model_len=32000
...
INFO 12-23 17:54:22 worker.py:241] the current vLLM instance can use total_gpu_memory (79.33GiB) x gpu_memory_utilization (0.60) = 47.60GiB
INFO 12-23 17:54:22 worker.py:241] model weights take 19.88GiB; non_torch_memory takes 0.11GiB; PyTorch activation peak memory takes 15.15GiB; the rest of the memory reserved for KV Cache is 12.47GiB.
```

In this case I asked to use 47.6GiB, but since we expect 15GiB to be used by the model when it runs we only end up allocating 33GiB

```
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.12             Driver Version: 535.104.12   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA A100-SXM4-80GB          On  | 00000000:08:01.0 Off |                    0 |
| N/A   36C    P0              78W / 400W |  33885MiB / 81920MiB |      0%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
```

**Comment by StrikerRUS** [comment]
@joerunde Thank you so much for your detailed response! Really appreciate it!

vllm is a great project and people behind it are doing just awesome work! There is no doubt that the current state of the project is a result of really hard work.

However, as a not well qualified user I'm expecting that software is working as it's described in it's docs. I totally understand that it's nearly impossible to guarantee that `gpu_memory_utilization` will work with a byte precision. I believe that a note in the docs clearly describing situations when vllm can go beyond assigned limits will be helpful for many users.

Regarding your points about my specific situation.

- cuda graphs - I don't think this is my case because as you can see from my docker compose I'm passing `--enforce-eager` flag
- fragmentation - not sure what you mean exactly, but guess it can be a memory leak cause
- memory that hasn't been GCed from running requests - perhaps the most likely reason because those extra 600MiB appears after some (random) number of requests
- memory that was used on the device outside of vLLM - it cannot be the reason, because only vllm is running on the host

Also, I strongly believe that those extra MiB are somehow related to the LoRA adapter usage, because there were no any problems with memory consumption before using adapters.

**Comment by joerunde** [comment]
> fragmentation - not sure what you mean exactly, but guess it can be a memory leak cause

I'm not talking about memory leaks though, what I mean here is that the pytorch allocator will reserve blocks of memory to allocate tensors within and as tensors are allocated and freed those blocks may become more fragmented. When you actually end up with an OOM error you can see some info about the memory that torch has reserved but not allocated, like this:

```
CUDA out of memory. Tried to allocate 524.00 MiB. GPU 0 has a total capacity of 15.77 GiB of which 137.12 MiB is free. Process 6513 has 15.63 GiB memory in use. Of the allocated memory 13.81 GiB is allocated by PyTorch, with 106.00 MiB allocated in private pools (e.g., CUDA Graphs), and 121.43 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation. See documentation for Memory Management (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)')
```

That shows that torch was holding onto 121MiB of memory that didn't actually have anything allocated in it, which is likely just fragmented space in vram. 

> Also, I strongly believe that those extra MiB are somehow related to the LoRA adapter usage

We do try to account for LoRA usage in the memory profiling by creating dummy adapters but it sounds very reasonable that using real adapters could cause more memory fragmentation issues, or lead to more things sitting around and not getting garbage collected for a while. I haven't looked into that for myself yet

**Comment by joe-schwartz-certara** [comment]
I've noticed that in these log lines:
`model weights take 37.10GiB; non_torch_memory takes 0.13GiB; PyTorch activation peak memory takes 22.66GiB; the rest of the memory reserved for KV Cache is 16.89GiB`
the activation peak memory scales linearly with max_model_len for the models I'm using. Is there a formula I can use to get this value before I launch vllm? It's nice to be able to predict what precise deployment hardware I will need before I even request it from the cloud for each model I am trying to launch in parallel. I have a system to calculate the kv-cache that always matches what that logline says for kv cache memory reservation when the model just barely has enough space to fit on the GPU resources I've given it (which is something I am always trying to do...)

**Comment by joerunde** [comment]
> Is there a formula I can use to get this value before I launch vllm? 

I'd love it if you figured one out!
At the moment the memory profiling is all empirical- we run the model and track the memory usage. There's no formula under the hood to calculate memory usage
